From 9e5dd10c9bc24ecd9f8a50c0cf53f849610121c1 Mon Sep 17 00:00:00 2001
From: Paulo Neves <ptsneves@gmail.com>
Date: Tue, 12 Oct 2021 11:32:15 +0200
Subject: [PATCH 1/1] [HSPP-6723] Set hostname in the kernel command line.

This is needed so that the hostname is passed to the kernel command
line. We do that by reading the gateway id to a u-boot env variable
which we later use
---
 arch/arm/mach-stm32mp/cpu.c     | 33 +++++++++++++++++++++++++++++++++
 drivers/fastboot/fb_getvar.c    | 29 +++++------------------------
 include/config_distro_bootcmd.h |  1 +
 3 files changed, 39 insertions(+), 24 deletions(-)

diff --git a/arch/arm/mach-stm32mp/cpu.c b/arch/arm/mach-stm32mp/cpu.c
index 4a5d0571a5..0a383dba0b 100644
--- a/arch/arm/mach-stm32mp/cpu.c
+++ b/arch/arm/mach-stm32mp/cpu.c
@@ -17,6 +17,7 @@
 #include <asm/arch/sys_proto.h>
 #include <dm/device.h>
 #include <dm/uclass.h>
+#include <fs.h>
 #include <linux/bitops.h>
 #include <power/pmic.h>
 
@@ -665,6 +666,37 @@ static int check_stsafe_provisioning(void)
 	return 0;
 }
 
+static int setup_gateway_id(void)
+{
+	char gateway_id[20] = {0};
+	loff_t fileSize = 0;
+	int ret;
+
+	ret = fs_set_blk_dev("mmc", "1:f", FS_TYPE_ANY);
+	if (ret == 0) {
+		ret = fs_size("/gwid/hostname", &fileSize);
+
+		if (ret == 0 && fileSize > sizeof(gateway_id) - 1)
+			ret = -EINVAL;
+	}
+
+	if (ret == 0)
+		ret = fs_set_blk_dev("mmc", "1:f", FS_TYPE_ANY);
+
+	if (ret == 0) {
+		loff_t actualRead = 0;
+		/* We do not read the last character because it is a newline character */
+		ret = fs_read("/gwid/hostname", (ulong)gateway_id, 0, fileSize-1, &actualRead);
+	}
+
+	if (ret == 0) {
+		printf("Read gateway id %s\n", gateway_id);
+		env_set("gateway_id", gateway_id);
+	}
+
+	return ret;
+}
+
 static int setup_wifi_mac_address(void)
 {
 	char envvar[26] = {0};
@@ -851,6 +883,7 @@ int arch_misc_init(void)
 	setup_boot_mode();
 	setup_mac_address();
 	setup_wifi_mac_address();
+	setup_gateway_id();
 	setup_serial_number();
 	get_longpress_ponkey();
 	get_reset_reason();
diff --git a/drivers/fastboot/fb_getvar.c b/drivers/fastboot/fb_getvar.c
index dc2f02525d..afab6aee84 100644
--- a/drivers/fastboot/fb_getvar.c
+++ b/drivers/fastboot/fb_getvar.c
@@ -265,31 +265,12 @@ static void getvar_is_userspace(char *var_parameter, char *response)
 
 static void getvar_gateway_id(char *gateway_id, char *response)
 {
-	char data[20] = {0};
-	loff_t fileSize = 0;
-	int ret;
+	const char *tmp = env_get("gateway_id");
 
-	ret = fs_set_blk_dev("mmc", "1:f", FS_TYPE_ANY);
-	if (ret == 0) {
-		ret = fs_size("/gwid/hostname", &fileSize);
-
-		if (ret == 0 && fileSize > sizeof(data) - 1)
-			ret = -EINVAL;
-	}
-
-	if (ret == 0)
-		ret = fs_set_blk_dev("mmc", "1:f", FS_TYPE_ANY);
-
-	if (ret == 0) {
-		loff_t actualRead = 0;
-		ret = fs_read("/gwid/hostname", (ulong)data, 0, fileSize, &actualRead);
-	}
-
-	if (ret == 0) {
-		fastboot_response("OKAY", response, "%s", data);
-	} else {
-		fastboot_fail("Failed to read gwid", response);
-	}
+	if (tmp)
+		fastboot_response("OKAY", response, "%s", tmp);
+	else
+		fastboot_fail("Value not set", response);
 }
 
 static void getvar_rtc(char *rtc, char *response)
diff --git a/include/config_distro_bootcmd.h b/include/config_distro_bootcmd.h
index 2ceb5a5df5..9aacebfac7 100644
--- a/include/config_distro_bootcmd.h
+++ b/include/config_distro_bootcmd.h
@@ -554,6 +554,7 @@
 			"${factory_reset} "                               \
 			"hw_conf=${hw_conf} "                             \
 			"rtc_vendor=${rtc_vendor} "                       \
+			"systemd.hostname=${gateway_id} "                 \
 			"${provision_stsafe} \0"                          \
 
 #ifndef CONFIG_BOOTCOMMAND
-- 
2.27.0
